### PR TITLE
Fine grained permissions for IdP, Realm Roles and Groups

### PR DESCRIPTION
**Pull request by xRodney**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes https://github.com/adorsys/keycloak-config-cli/issues/663

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: src/main/java/de/adorsys/keycloak/config/resource/ManagementPermissions.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+import javax.ws.rs.Consumes;
+import javax.ws.rs.GET;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+
+/**
+ * Routes which are somehow missing from the official Keycloak client library
```
Comment (jkroepke) [inline src/main/java/de/adorsys/keycloak/config/resource/ManagementPermissions.java:14]: https://github.com/keycloak/keycloak/pull/10724

File: src/main/java/de/adorsys/keycloak/config/service/clientauthorization/GroupPermissionResolver.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        this.realmName = realmName;
+        this.groupRepository = groupRepository;
+    }
+
+    @Override
+    public String resolveObjectId(String groupPath, String authzName) {
+        try {
+            return groupRepository.getGroupByPath(realmName, groupPath).getId();
+        } catch (NotFoundException | KeycloakRepositoryException e) {
+            throw new ImportProcessingException("Cannot find group with path '%s' in realm '%s' for '%s'", groupPath, realmName, authzName);
```
Comment (xRodney) [inline src/main/java/de/adorsys/keycloak/config/service/clientauthorization/GroupPermissionResolver.java:47]: The negative scenarios like these are currently not covered by tests. Should they be?

Comment (jkroepke) [reply src/main/java/de/adorsys/keycloak/config/service/clientauthorization/GroupPermissionResolver.java:47]: In history, it was worth to have tests to tests failures. Thats would be great, to have some tests here, too.

From what I now, you have to add a typo to the resource which be result into a group/client/role/idp not found.

Comment (xRodney) [reply src/main/java/de/adorsys/keycloak/config/service/clientauthorization/GroupPermissionResolver.java:47]: Okay, added a few

File: src/test/java/de/adorsys/keycloak/config/service/ImportClientsIT.java [hunk 3]
```diff
@@ -1971,6 +1996,342 @@ void shouldRemoveAuthzPoliciesForRealmManagement() throws IOException {
...
+    @Test
+    @Order(50)
+    void shouldNotTriggerErrorWhenReferencingInvalidUuidInFineGrainedAuthz() throws IOException {
+        // These scenarios do not use placeholders and instead reference objects by UUID - which do not need to exist.
+        // Keycloak accepts this and it sometimes even works (for objects that allow specifying UUID in creation and are created after the import)
+        // This is how to partially support fine-grained authz for types that are not supported yet by keycloak-config-cli
+        // keycloak-config-cli will log a warning, but otherwise the import succeeds
+
+        RealmImport foundImport0 = getFirstImport("50.0_update_realm_update_authz_policy_for_unknown_type_with_id_realm-management.json");
+        realmImportService.doImport(foundImport0);
```
Comment (jkroepke) [inline src/test/java/de/adorsys/keycloak/config/service/ImportClientsIT.java:2320]: Let's have some asserts here. 

```suggestion
        assertDoesNotThrow(() -> realmImportService.doImport(foundImport0));
```

File: src/test/java/de/adorsys/keycloak/config/service/ImportClientsIT.java [hunk 4]
```diff
@@ -1971,6 +1996,342 @@ void shouldRemoveAuthzPoliciesForRealmManagement() throws IOException {
...
+
+        RealmImport foundImport1 = getFirstImport("50.1_update_realm_update_authz_policy_for_client_with_bad_id_realm-management.json");
+        realmImportService.doImport(foundImport1);
```
Comment (jkroepke) [inline src/test/java/de/adorsys/keycloak/config/service/ImportClientsIT.java:2323]:
```suggestion
        assertDoesNotThrow(() -> realmImportService.doImport(foundImport1));
```

File: src/test/java/de/adorsys/keycloak/config/service/ImportClientsIT.java [hunk 5]
```diff
@@ -1971,6 +1996,342 @@ void shouldRemoveAuthzPoliciesForRealmManagement() throws IOException {
...
+
+        RealmImport foundImport2 = getFirstImport("50.2_update_realm_update_authz_policy_for_idp_with_bad_id_realm-management.json");
+        realmImportService.doImport(foundImport2);
```
Comment (jkroepke) [inline src/test/java/de/adorsys/keycloak/config/service/ImportClientsIT.java:2326]:
```suggestion
        assertDoesNotThrow(() -> realmImportService.doImport(foundImport2));
```

File: src/test/java/de/adorsys/keycloak/config/service/ImportClientsIT.java [hunk 6]
```diff
@@ -1971,6 +1996,342 @@ void shouldRemoveAuthzPoliciesForRealmManagement() throws IOException {
...
+
+        RealmImport foundImport3 = getFirstImport("50.3_update_realm_update_authz_policy_for_role_with_bad_id_realm-management.json");
+        realmImportService.doImport(foundImport3);
```
Comment (jkroepke) [inline src/test/java/de/adorsys/keycloak/config/service/ImportClientsIT.java:2329]:
```suggestion
        assertDoesNotThrow(() -> realmImportService.doImport(foundImport3));
```

File: src/test/java/de/adorsys/keycloak/config/service/ImportClientsIT.java [hunk 7]
```diff
@@ -1971,6 +1996,342 @@ void shouldRemoveAuthzPoliciesForRealmManagement() throws IOException {
...
+
+        RealmImport foundImport4 = getFirstImport("50.4_update_realm_update_authz_policy_for_group_with_bad_id_realm-management.json");
+        realmImportService.doImport(foundImport4);
```
Comment (jkroepke) [inline src/test/java/de/adorsys/keycloak/config/service/ImportClientsIT.java:2332]:
```suggestion
        assertDoesNotThrow(() -> realmImportService.doImport(foundImport4));

```

**Review by jkroepke — COMMENTED** [review]
Hi @xRodney 

through you recently added test, I saw some stacktraces in the junit output with are uncommon for me.

I could verify that keycloak-config-cli is raise a stack track without any reason to do it.
```

% java -jar target/keycloak-config-cli.jar --keycloak.url=http://localhost:8080 --keycloak.password=admin123 --import.path=./src/test/resources/import-files/clients/50.0_update_realm_update_authz_policy_for_unknown_type_with_id_realm-management.json
2022-03-16 08:31:35.012  INFO 8141 --- [           main] d.a.k.config.KeycloakConfigApplication   : Starting KeycloakConfigApplication v4.8.2-SNAPSHOT using Java 11.0.14.1 on joe-nb with PID 8141 (/Users/jkr/IdeaProjects/private/keycloak-config-cli/target/keycloak-config-cli.jar started by jkr in /Users/jkr/IdeaProjects/private/keycloak-config-cli)
2022-03-16 08:31:35.016  INFO 8141 --- [           main] d.a.k.config.KeycloakConfigApplication   : No active profile set, falling back to 1 default profile: "default"
2022-03-16 08:31:35.516  INFO 8141 --- [           main] d.a.k.config.KeycloakConfigApplication   : Started KeycloakConfigApplication in 1.019 seconds (JVM running for 1.407)
2022-03-16 08:31:36.233  INFO 8141 --- [           main] d.a.k.c.provider.KeycloakImportProvider  : 1 configuration files found.
2022-03-16 08:31:36.235  INFO 8141 --- [           main] d.a.k.c.provider.KeycloakImportProvider  : Loading file '/Users/jkr/IdeaProjects/private/keycloak-config-cli/src/test/resources/import-files/clients/50.0_update_realm_update_authz_policy_for_unknown_type_with_id_realm-management.json'
2022-03-16 08:31:36.354  INFO 8141 --- [           main] d.a.k.config.KeycloakConfigRunner        : Importing file '/Users/jkr/IdeaProjects/private/keycloak-config-cli/src/test/resources/import-files/clients/50.0_update_realm_update_authz_policy_for_unknown_type_with_id_realm-management.json'
2022-03-16 08:31:39.574  WARN 8141 --- [           main] a.k.c.s.ClientAuthorizationImportService : Unable to enable permissions for 'unknowntype.resource.00000000-00000000-00000000-00000000' in realm 'realmWithClientsForAuthzGrantedPolicies'. Import will continue, but may fail later.

de.adorsys.keycloak.config.exception.ImportProcessingException: Cannot resolve 'unknowntype.resource.00000000-00000000-00000000-00000000' in realm 'realmWithClientsForAuthzGrantedPolicies', the type 'unknowntype' is not supported by keycloak-config-cli.
        at de.adorsys.keycloak.config.service.ClientAuthorizationImportService$RealmManagementPermissionsResolver.getPermissionResolver(ClientAuthorizationImportService.java:594)
        at de.adorsys.keycloak.config.service.ClientAuthorizationImportService$RealmManagementPermissionsResolver.enableFineGrainedPermission(ClientAuthorizationImportService.java:583)
        at de.adorsys.keycloak.config.service.ClientAuthorizationImportService$RealmManagementPermissionsResolver.createFineGrantedPermissions(ClientAuthorizationImportService.java:566)
        at de.adorsys.keycloak.config.service.ClientAuthorizationImportService.updateAuthorization(ClientAuthorizationImportService.java:134)
        at de.adorsys.keycloak.config.service.ClientAuthorizationImportService.updateClientAuthorizationSettings(ClientAuthorizationImportService.java:115)
        at de.adorsys.keycloak.config.service.ClientAuthorizationImportService.doImport(ClientAuthorizationImportService.java:92)
        at de.adorsys.keycloak.config.service.RealmImportService.configureRealm(RealmImportService.java:202)
        at de.adorsys.keycloak.config.service.RealmImportService.createRealm(RealmImportService.java:171)
        at de.adorsys.keycloak.config.service.RealmImportService.doImport(RealmImportService.java:136)
        at de.adorsys.keycloak.config.KeycloakConfigRunner.run(KeycloakConfigRunner.java:73)
        at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:777)
        at org.springframework.boot.SpringApplication.callRunners(SpringApplication.java:761)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:310)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1312)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1301)
        at de.adorsys.keycloak.config.KeycloakConfigApplication.main(KeycloakConfigApplication.java:35)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:49)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:108)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:58)
        at org.springframework.boot.loader.PropertiesLauncher.main(PropertiesLauncher.java:467)

2022-03-16 08:31:40.286  INFO 8141 --- [           main] d.a.k.config.KeycloakConfigRunner        : keycloak-config-cli running in 00:04.056.
```

This may confuse people, because by reading the output they don't know if the error is mentionable or not. At least such users will create an issue here and ask if the behavior is okay.

**Comment by xRodney** [comment]
I've removed the stack traces from logs and made the warning nicer.

Examples:
```
2022-03-16 09:53:11.223  WARN   --- [           main] a.k.c.s.ClientAuthorizationImportService : Unable to enable permissions for 'unknowntype.resource.00000000-00000000-00000000-00000000'. Import will continue, but may fail later. Reason: Cannot resolve 'unknowntype.resource.00000000-00000000-00000000-00000000' in realm 'realmWithClientsForAuthzGrantedPolicies', the type 'unknowntype' is not supported by keycloak-config-cli.
2022-03-16 09:53:11.711  WARN   --- [           main] a.k.c.s.ClientAuthorizationImportService : Unable to enable permissions for 'client.resource.00000000-00000000-00000000-00000000'. Import will continue, but may fail later. Reason: Cannot find client with id '00000000-00000000-00000000-00000000' in realm 'realmWithClientsForAuthzGrantedPolicies'
2022-03-16 09:53:12.254  WARN   --- [           main] a.k.c.s.ClientAuthorizationImportService : Unable to enable permissions for 'idp.resource.00000000-00000000-00000000-00000000'. Import will continue, but may fail later. Reason: Cannot find identity provider with internal id '00000000-00000000-00000000-00000000' in realm 'realmWithClientsForAuthzGrantedPolicies'
2022-03-16 09:53:12.811  WARN   --- [           main] a.k.c.s.ClientAuthorizationImportService : Unable to enable permissions for 'role.resource.00000000-00000000-00000000-00000000'. Import will continue, but may fail later. Reason: Cannot find realm role with id '00000000-00000000-00000000-00000000' in realm 'realmWithClientsForAuthzGrantedPolicies'
2022-03-16 09:53:13.347  WARN   --- [           main] a.k.c.s.ClientAuthorizationImportService : Unable to enable permissions for 'group.resource.00000000-00000000-00000000-00000000'. Import will continue, but may fail later. Reason: Cannot find group with id '00000000-00000000-00000000-00000000' in realm 'realmWithClientsForAuthzGrantedPolicies'
```

**Comment by jkroepke** [comment]
Hi @xRodney 

thanks a lot for you awesome work. Huge contribution.

**Comment by xRodney** [comment]
Always a pleasure, many thanks for your awesome project @jkroepke :D

**Pull request merged**